### ISSUE TITLE
1.16.1: Custom Javascript issues.  

### ISSUE THREAD

**Issue by mrcoffee1026** [kangarko/ChatControl-Pro#1805]
Javascript variables are still not being resolved.  The Javascript IS being processed, I can tells since it makes any multiline value completely illegible after the plugin loads once.  {ping} always prints 0.. and I'm not even sure where it's getting the 0 since I can't get any other function to so much as return a single character.   If a change is made to any of the javascript files, a server restart is needed to test them, chc reload does NOT reload js.

**Comment by mrcoffee1026** [comment]
Ahh nevermind about the illegible part... must have been a mistake in one of them or something it usually leaves it alone otherwise, but I still can't actually use values in formatting.yml later, nothing is ever returned from the scripts.

**Comment by AngellitoX** [comment]
Hello, can you check if this is similar to #1799 ?

There has been some issues reported on PlaceholderAPI where Matej is still waiting to receive some help from their side.

Angel from MineAcademy.org

**Comment by mrcoffee1026** [comment]
No, not similar.  The placeholder problem predates 1.15.2 and doesn't have anything to do with ChatControl.  They claim it's because nashore isn't included in my version of java because it's deprecated... but it has never been included in anything other that the JDK and that's not what you install to run minecraft.  Javascript still works absolutely fine in chatcontrol if using 1.15.2 and a version of chatcontrol before the introduction of version 9, so it is NOT the version of java that is the issue, but I have heard that it maybe possible to make js work in PlaceHolderAPI if you install something like Java 11, but I'm not sure if I'm remembering the version someone had said had worked for them.

**Comment by kangarko** [comment]
Can you try running Java JDK 8 please?

**Comment by mrcoffee1026** [comment]
Doesn't seem to be a way to do it: I have attempted everything suggested for doing this, and versin still returns the JRE info.  A JRE is absolutely essential for running minecraft as far as I know.  In any case, even it did solve the PlaceHolderAPI problem, that isn't the problem I'm having, as I already explained.  Keep in mind that Javascript works fine in ChatControl 8.6.6 (before the changes post version 9) in 1.15.2... and the PlaceHolderAPI javascript problem existed at that time as well.  Only thing is, I have no idea when they modified PlaceHolderAPI and made it stop working in Java 8 since from for all this time I've been using Chatcontrol to do javascript since it was easier and did not - at that time - need a restart to try changes in the script.

**Comment by mrcoffee1026** [comment]
OK, my bad, I maybe forgot a placeholder so my if statements were always false and THAT is why I never got a value returned.  I think the ping is always 0 since I'm testing on the local machine but I can't be certain.  However, the javascript file IS modified on server start,  Can you make it NOT do that?  It screws up the formatting and just makes it difficult to read or do anything if a change is needed.

**Comment by mrcoffee1026** [comment]
Have also found the instead of rendering a \n as a new line like it should, formatting.yml simply prints the "\n".

**Comment by AngellitoX** [comment]
Hey coffee! Would you mind creating a separate ticket for your last discovery? In advance thank you wo much for the heads up!

**Comment by mrcoffee1026** [comment]
I did that.. and this is just a more general question: under what circumstances would that example item.yml javascript hover item ever be useable?  I've tried to use do something like use a rule and a handler to trigger it or even stick the {item} placeholder directly into formatting.yml but there never seems to be any condition under which the hover works... I assume there must be some condition under which the hover works or "Hover_item:" would not be a thing... so what exactly is the necessary condition for when this actually works?

**Comment by kangarko** [comment]
Thanks, I do believe it's related to PlaceholderAPI, can you try uninstalling this plugin for a test round to see if our JavaScript placeholders will start to work?

I am not aware of the lines being crunched up like this, sorry I cannot reproduce. Possibly it can be MultiCraft or your server panel.

The javascript hover is planned feature for 9.0.0

**Comment by mrcoffee1026** [comment]
No, as I explained before, there is a PlaceholderAPI javascript issue which is completely unrelated to ChatControl.  Nothing you do in your plugin will fix theirs, nothing they do in their plugin will affect yours.  Earlier in this ticket I had found the javascript and placeholders to be operational short of some unicode issues.

The code smashing being done on start is not being done by MultiCraft since I do not have it.  This is being run on an independent unix server from the command line in a screen wrapper, there is no other fancy control panel modifying code on launch.

I will look forward to your javascript hover feature, we really want to use that item example script to work for something like players typing in [item] in chat to show what cool tool they have.

**Comment by kangarko** [comment]
Thanks.

I apology for not being able to help with the line smashing. If you find steps to reproduce other than simply editing out the file and reloading, please do let me know below.

However there _is_ a problem with PlaceholderAPI that we are currently monitoring at https://github.com/kangarko/ChatControl-Pro/issues/1799

**Comment by mrcoffee1026** [comment]
Yeah I'm not sure I can give you the steps since it doesn't ALWAYS happen for me either, just sometimes it does that.  And yeah I can't help with the placeholder thing I just know they're doing something wrongly since - from what I can tell - chatcontrol ALSO uses the nashorn engine and has no trouble with it.  For the time being I have simply disabled the javascript extension in Placeholders since I'm not using theirs for anything.

**Comment by kangarko** [comment]
Got it. Sorry to hear you had to disable them, I hope extendedclip will get back to this soon. I already found the fix.

**Comment by mrcoffee1026** [comment]
Well I haven't been using their javascript since I found I could do it in ChatControl and - at the time - only had to /chc reload instead of restart the server to see the changes.  Seemed to me I had to restart the server for any changes made to a papi javascript so it just seemed inconvenient.  But it was tossing some error about the engine in 1.15.2 and so I don't know when it started doing that as I hadn't been using js in there since... I don't know... probably 1.12... so there's no way of knowing - for me - when that error started showing, I just know it didn't have anything to do with the version of Java.
